Introduction and goals/non-goals (#4)
jakartaee/query: spec/src/main/asciidoc/chapters/introduction/introduction.asciidoc

== Introduction

:Jakarta-Persistence: https://jakarta.ee/specifications/persistence/
:Jakarta-Data: https://jakarta.ee/specifications/data/
:Jakarta-NoSQL: https://jakarta.ee/specifications/nosql/

Jakarta Query defines an object-oriented query language designed for use with
link:{Jakarta-Persistence}[Jakarta Persistence], link:{Jakarta-Data}[Jakarta Data],
and link:{Jakarta-NoSQL}[Jakarta NoSQL], with:

> - a core language which can be implemented by Jakarta Data and Jakarta NoSQL
> providers using non-relational datastores, and

Reviewer: It's probably best not to limit the Jakarta Data case for the core language to non-relational because a Jakarta Data provider for relational databases might use this query language for an implementation based on JDBC or SQL.  I don't know if any providers will actually do that, but it's possible and should be allowed for if anyone wanted to.  I suppose another possibility would be a Data provider based on Jakarta Persistence, but choosing to limit the user to the core query language.
```suggestion
- a core language which can be implemented by Jakarta Data and Jakarta NoSQL
providers, and
```

Jakarta Query prioritizes clients written in Java. However, it is not by
nature limited to Java, and implementations in other sufficiently Java-like
programming languages are encouraged.

=== Object-oriented query languages

A data structure in an object-oriented language is a graph of objects
interconnected by unidirectional object references, which may be polymorphic.
Some non-relational databases support similar representations. On the other
hand, relational databases represent relationships between entities using
foreign keys, and therefore SQL has no syntactic construct representing
navigation of an association. Similarly, inheritance and polymorphism can be
easily represented within the relational model, but are not present as
first-class constructs in the SQL language. An object-oriented query language
is a dialect of SQL with support for associations and subtype polymorphism.

=== Historical background

:JPQL: https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#a4665
:JDQL: https://jakarta.ee/specifications/data/1.0/jakarta-data-1.0#_jakarta_data_query_language
:JSR-220: https://jcp.org/en/jsr/detail?id=220

> Object-oriented dialects of SQL have existed since at least the early 90s.

Reviewer: ```suggestion
Object-oriented dialects of SQL have existed since at least the early 1990s.
```

The Object Query Language (OQL) was an early example, targeting object
databases, but was never widely used, since object databases were themselves
not widely adopted. Hibernate Query Language (HQL) and the Enterprise JavaBeans
Query Language (EJB-QL) were both introduced in 2001 as query languages
intended for use with object/relational mapping. HQL was widely adopted by the
Java community and was eventually standardized as the
link:{JPQL}[Java Persistence Query Language] (JPQL) by link:{JSR-220}[JSR-220]
in 2006. JPQL has been implemented by at least five different products and is
in extremely wide use today. On the other hand, since JPQL is defined as part
of the Jakarta Persistence specification, it has not been reused outside the

> context of object/relational mapping in Java. More recently, Jakarta Data 1.0

Reviewer: To allow for when 2024 is no longer recent,
```suggestion
context of object/relational mapping in Java. In 2024, Jakarta Data 1.0
```
